Documentation/quickstart/compiling.rst

```diff
@@ -1,57 +1,82 @@
+       
+The output is in the format ``<board name>:<board configuration>``. You will see that
+generally all boards support the ``nsh`` configuration which is a good sarting point
+since it enables booting into the interactive command line
+:doc:`/components/nsh/index`.
 
-#. Initialize Configuration
-
-   Pick one of the board:application base configuration pairs from the list, and feed it to the
-   configuration script. The ``-l`` tells use that we're on Linux. macOS and Windows builds are
-   possible, this guide doesn't cover them yet.
+To choose a configuration you pass the ``<board name>:<board configuration>`` option
+to ``configure.sh`` and indicate your host platform, such as:
 
-    .. code-block:: bash
+    .. code-block:: console
 
        $ cd nuttx
-       $ # this is the basic layout of the command:
-       $ # ./tools/configure.sh -l <board-name>:<config-dir>
-       $ # for example:
-       $ ./tools/configure.sh -l sama5d2-xult:nsh
+       $ ./tools/configure.sh -l stm32f4discovery:nsh
+       
+The ``-l`` tells use that we're on Linux (macOS and Windows builds are
+possible). Use the ``-h`` argument to see all available options.
+
+Customize Your Configuration (Optional)
+=======================================
 
-#. Customize Your Configuration (Optional)
+This step is optional. Right now, this is mainly to get familiar with how it
+works– you don't need to change any of the options now, but knowing how
+to do this will come in handy later.
 
-   This step is optional. Right now, this is mainly to get familiar with how it works– you don't need to change
-   any of the options now, but knowing how to do this will come in handy later.
+There are a lot of options. We'll cover a few of them here.
+Don't worry about the complexity– you don't have to use most of the options.
 
-   There are a lot of options. We'll cover a few of them here. Don't worry about the complexity– you don't have to use most of the options.
+.. code-block:: console
 
-    .. code-block:: bash
+   $ cd nuttx/
+   $ make menuconfig
+   
+.. todo::
+  Explain some useful options.
 
-       $ make menuconfig
+Build NuttX
+===========
 
-#. Compile NuttX
+We can now build NuttX. To do so, you can simply run:
 
-    .. code-block:: bash
+  .. code-block:: console
 
-       $ make clean; make
+     $ cd nuttx/
+     $ make make
+     
+The build will complete by generating the binary outputs
+inside `nuttx` directory. Typically this includes the `nuttx`
+ELF file (suitable for debugging using `gdb`) and a `nuttx.bin`
+file that can be flashed to the board.     
+     
+To clean the build, you can do:
 
-#. Install the Executable Program on Your Board
+  .. code-block:: console
 
-   This step is a bit more complicated, depending on your board. It's covered in the section
-   :ref:`Running Apache NuttX <running>`.
+     $ make clean
+     
+It is recommended that after manually modifying the configuration you first clean
+before building.
```

Comment:
Are we recommending somewhere to change the configuration file manually?
We've always been against that.  It's easy to mess things up that way.

Reply:
"manually" means going into menuconfig in this context.

Reply:
Ah, okay.  My head went elsewhere.
Did you run into a problem where you needed to clean after changing a config with menuconfig?

Reply:
Yes, I think we should address this eventually. I found that sometimes an old .o remains and when building it gets linked as well. It is difficult to reproduce so I always put it off when it happens.

Reply:
We've been talking about some similar corner cases around cleaning/distcleaning not long ago.
Yes, they are all hard to reproduce, it would've been great if we could, then we can address them and eliminate the issue.
But, of course, this is out of the scope of this PR.

Reply:
I will play around in a while to see if I can reproduce it. Do you prefer to leave the "make clean" instruction out?

Reply:
I wanted to.  But this isn't the first time I hear about these corner cases.
I personally rarely "clean" after a "menuconfig".
I guess we can keep it for now, and get back to it after some testing.

Reply:
Ok, I've found a case, I will open an issue to discuss it. Anyway, I can open a PR again to remove the "make clean" if needed so no need to hold this for that.

Reply:
If the user modify defconfig, it's better to make a full build with the current build system, otherwise the strange error happen either at build time or run time. Here is a partial list I have saw so far:

1. Use the wrong compile option if user change CONFIG_DEBUG_OPTLEVEL and related option
2. The old *.bdat and *.pdat left under apps/builtin/registy/
3. And issue report here: https://github.com/apache/incubator-nuttx/issues/1764

These issues are hard to debug especially for the newie who doesn't know the flaw inside build system. I would suggest to force the full build if config.h or .config has been changed before we fix all known issues.

Reply:
But as you mention these are bugs (in fact, now that you point them out, I will try to address them as well). I don't really like to convey that NuttX simply requires the `make clean` under normal circumstances. I think, considering the situation, I would change this to be a "note" saying something like "since there are currently some issues in NuttX build system, please do make clean after reconfigure". That way it is clear this is not intentional. What do you think?

Reply:
Ok, I just added this change.

Reply:
"The old *.bdat and *.pdat left under apps/builtin/registy/" is this still an issue?  If I remember correctly this issue lived for a very short time due to a change in the part that generates/cleans those files, but then got fixed rather quickly.

Reply:
I am not sure whether all issues get resolved. The build system stop working mysteriously(either build time or run time) after the user make the change through menuconfig which is the top most issue reported in our internal forum.

Reply:
I'm looking at this right now. It shouldn't be a problem since I see that the *.bdat actually depends on .config. Anyway, if you ever come across this issue again please open an issue.

Reply:
> Use the wrong compile option if user change CONFIG_DEBUG_OPTLEVEL and related option

This is the most problematic. I just verified this is a problem. Any change in flags should trigger a rebuild of files whose flags have changed (the simple approach would be to consider every file affected). I'll open an issue for this.

Reply:
@xiaoxiang781216 I guess it was introduced and fixed by #1072.  I know it's weird :P but if you had it internally for some time you could've faced the issue.

Reply:
Please resolve this conversation if you feel it look OK now